### PR TITLE
[PHP] Multiple default namespace navigator fix, FixImports for the whole file

### PR DESCRIPTION
**Pull request by rossluk**
This PR fixes next issues:

1. Incorrect navigator view when you have bracketed default namespace or multiple default namespaces in file:

-- before fix it shows classes without namespace, and USE from second default namespace is showed as first one
navigator_view

-- after fix
navigator_view_fix

2. Incorrect fix imports when there are multiple default namespaces and some of them has imported USES:

-- initial state
getns_from_other

-- before fix, it gets all uses from all default namespaces and inserts it in the first one
get_ns_from_other_after
-- after fix 
getns_from_other_fix

3. File state change after each fix import when there is no changes

-- before fix before fix import action (see tab)
file_change

-- before fix after fix import action (see tab)
file_change_after

Also with the fixes I've changed Fix Imports logic. Previously when you use Fix Imports outside bracketed namespace it was changing first available namespace uses, but now if you in global scope it fixes all the file, see below:

-- initial state
Screen Shot 2023-03-19 at 15 40 04
-- after Fix Imports
Screen Shot 2023-03-19 at 13 08 23

To have Fix Imports working for the whole file I had to add scope to UsedNamesCollector's names and it also affected some tests, because for now it is important that DataItems from ImportDataCreator were paired with Selections, but previously Selection could be imported without real DataItem.

### CODE REVIEW

File: php/php.editor/src/org/netbeans/modules/php/editor/actions/ImportDataCreator.java [hunk 1]
```diff
@@ -203,7 +219,7 @@ private boolean hasDefaultNamespaceName(final Collection<FullyQualifiedElement>
     }
 
     private interface PossibleItem {
-
+
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/actions/ImportDataCreator.java:222]: Please remove.

File: php/php.editor/src/org/netbeans/modules/php/editor/model/ModelUtils.java [hunk 2]
```diff
@@ -425,7 +438,7 @@ public static NamespaceScope getNamespaceScope(FileScope fileScope, int offset)
         NamespaceScope retval = fileScope.getDefaultDeclaredNamespace();
         Collection<? extends NamespaceScope> declaredNamespaces = fileScope.getDeclaredNamespaces();
         for (NamespaceScope namespaceScope : declaredNamespaces) {
-            OffsetRange blockRange = namespaceScope.getBlockRange();
+            OffsetRange blockRange = namespaceScope.getBlockRange();
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/model/ModelUtils.java:441]: Please remove trailing spaces.

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses [hunk 3]
```diff
@@ -23,8 +23,6 @@ declare (strict_types=1);
 
 namespace TestNamespace;
 
-use Foo\Foo\Foo;
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses:26]: Why?

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/02/testNB4978_02.php [hunk 4]
```diff
@@ -35,3 +35,7 @@ public function test(?Foo $foo): ?Foo
     }
 
 }
+
+namespace Foo\Foo;
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/actions/testNB4978/02/testNB4978_02.php:39]: Why?

**Comment by junichi11** [comment]
@rossluk This PR contains some fixes as one commit at the same time. So, you should separate commits(or separate to PRs).

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses (hunk 3 above)
Comment (junichi11) [reply php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses:26]: > To have Fix Imports working for the whole file I had to add scope to UsedNamesCollector's names and it also affected some tests, because for now it is important that DataItems from ImportDataCreator were paired with Selections, but previously Selection could be imported without real DataItem.

We fixed this once, so, it would be recognized as a regression, I think...

Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses:26]: It because there is no one real namespace in file or near to it, so if you debug UsesNamespaceCollector all dataItems names will be empty. So if I remove check for UsedNames you could put to selection anything like "Just\Want\It\To\Be\Added\To\Uses" and it would be added. For this test it would work as previously if we put file with \Foo\Foo\Foo near to the test file.

File: php/php.editor/src/org/netbeans/modules/php/editor/model/ModelUtils.java (hunk 2 above)
Comment (rossluk) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/ModelUtils.java]: Done

File: php/php.editor/src/org/netbeans/modules/php/editor/actions/ImportDataCreator.java (hunk 1 above)
Comment (rossluk) [reply php/php.editor/src/org/netbeans/modules/php/editor/actions/ImportDataCreator.java]: Done

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses (hunk 3 above)
Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/testNB4978_01.php.fixUses]: I've added class file to have tests in previous state and don't confuse anyone.

File: php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ModelBuilder.java [hunk 5]
```diff
@@ -86,7 +86,7 @@ class ModelBuilder {
     NamespaceScope build(NamespaceDeclaration node, OccurenceBuilder occurencesBuilder) {
         final NamespaceDeclarationInfo info = NamespaceDeclarationInfo.create(node);
 
-        NamespaceScopeImpl nScope = (info.isDefaultNamespace()) ? defaultNamespaceScope : ModelElementFactory.create(info, this);
+        NamespaceScopeImpl nScope = /*(info.isDefaultNamespace()) ? defaultNamespaceScope :*/ ModelElementFactory.create(info, this);
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ModelBuilder.java:89]: What is the reason for the comment-out?

File: php/php.editor/test/unit/data/testfiles/structure/php53/bracketed_multiple_namespaces_with_multiple_default.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/structure/php53/bracketed_multiple_namespaces_with_multiple_default.php:1]: Please remove this and add the license header.

File: php/php.editor/test/unit/data/testfiles/structure/php53/multiple_namespaces_with_elements_from_default.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+
+<?php
+class Connection {}
+namespace MyProject;
+class Connection {}
+function connect() {}
+namespace AnotherProject;
+class Connection {}
+function connect() {}
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/structure/php53/multiple_namespaces_with_elements_from_default.php:9]: Note: NB doesn't handle this as an error but it is fatal error. We cannot write classes before the Namespace declaration.

File: php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ModelBuilder.java (hunk 5 above)
Comment (rossluk) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ModelBuilder.java:89]: This is actually the fix for multiple default namespaces in navigator and for proper default namespaces OffsetRange. I left it commented only if in a near feature something will go wrong with this to give a clue to someone who will debug this. Should I remove the comment or do you need more deep explanation?

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/Foo.php [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<?php
+
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/Foo.php:2]: License header.

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/02/Foo.php [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<?php
+
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/actions/testNB4978/02/Foo.php:2]: License header.

File: php/php.editor/test/unit/data/testfiles/structure/php53/multiple_namespaces_with_elements_from_default.php (hunk 7 above)
Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/structure/php53/multiple_namespaces_with_elements_from_default.php:9]: Sorry, my bad. I've removed the test.

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/actions/FixUsesPerformerTest.java [hunk 10]
```diff
@@ -401,6 +401,24 @@ public void testGH5578_08() throws Exception {
...
+        String[] selections = new String[] {"\\NS2\\D"};
+        Options options = new Options(false, false, false, false, true);
+        performTest("// test^", createSelections(selections, ItemVariant.Type.CLASS), true, options);
+    }
+    
+    public void testWholeFileUse() throws Exception {
+        String[] selections = new String[] {"\\NS2\\B", "\\NS2\\D"};
+        Options options = new Options(false, false, false, false, true);
+        performTest("// test^", createSelections(selections, ItemVariant.Type.CLASS), true, options);
+    }
```
Comment (junichi11) [inline php/php.editor/test/unit/src/org/netbeans/modules/php/editor/actions/FixUsesPerformerTest.java:421]: Can you add a test case for trait?

File: php/php.editor/test/unit/data/testfiles/structure/php53/bracketed_multiple_namespaces_with_multiple_default.php (hunk 6 above)
Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/structure/php53/bracketed_multiple_namespaces_with_multiple_default.php]: Done

File: php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ModelBuilder.java (hunk 5 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/impl/ModelBuilder.java:89]: If you want it commented out, please, leave a comment as the reason because people other than you don't know why. Thanks!

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/Foo.php (hunk 8 above)
Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/actions/testNB4978/01/Foo.php:20]: Fixed

File: php/php.editor/test/unit/data/testfiles/actions/testNB4978/02/Foo.php (hunk 9 above)
Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/actions/testNB4978/02/Foo.php:20]: Fixed

File: php/php.editor/test/unit/src/org/netbeans/modules/php/editor/actions/FixUsesPerformerTest.java (hunk 10 above)
Comment (rossluk) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/actions/FixUsesPerformerTest.java:421]: Done

Comment (junichi11) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/actions/FixUsesPerformerTest.java:421]: I meant the case that `use NS2\F;`(trait) is inserted. Sorry for not being clear...

Comment (rossluk) [reply php/php.editor/test/unit/src/org/netbeans/modules/php/editor/actions/FixUsesPerformerTest.java:421]: Hope I understood it right.

File: php/php.editor/test/unit/data/testfiles/actions/testWholeFileUse/02/testWholeFileUse_02.php.fixUses [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// test
+namespace {
+
+use NS2\B;
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/actions/testWholeFileUse/02/testWholeFileUse_02.php.fixUses:24]: This case is added the class.
I meant...`use NS2\F;` here

Comment (rossluk) [reply php/php.editor/test/unit/data/testfiles/actions/testWholeFileUse/02/testWholeFileUse_02.php.fixUses]: If this time it is not right, give me please example block, thank you.

Comment (junichi11) [reply php/php.editor/test/unit/data/testfiles/actions/testWholeFileUse/02/testWholeFileUse_02.php.fixUses]: Thanks!

**Comment by junichi11** [comment]
Removing trailing whitespaces: It's OK only with your changed parts. (because If there are many changes for that, it makes it harder to review :)) Thanks!

**Review by junichi11 — APPROVED** [review]
Thank you for your work.

File: php/php.editor/src/org/netbeans/modules/php/editor/actions/FixUsesPerformer.java [hunk 12]
```diff
@@ -570,7 +614,7 @@ public void visit(DeclareStatement node) {
             if (CancelSupport.getDefault().isCancelled()) {
                 return;
             }
-            declareStatements.add(node);
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/actions/FixUsesPerformer.java:573]: As @junichi11 wrote, please, avoid changes in unrelated files and lines. The reason is that the Git history is then more difficult to read/follow, thanks. I think that the IDE can be configured to format/change only the changed lines (see IDE Options dialog).

File: php/php.editor/src/org/netbeans/modules/php/editor/actions/UsedNamesCollector.java [hunk 13]
```diff
@@ -143,10 +106,14 @@ private boolean existsUseForTypeName(final Collection<? extends UseScope> declar
 
     private static class NamespaceNameVisitor extends DefaultVisitor {
         private final OffsetRange offsetRange;
-        private final Map<String, List<UsedNamespaceName>> existingNames = new HashMap<>();
+        private final NamespaceScope namespaceScope;
+        private NamespaceScope currentScope;
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/actions/UsedNamesCollector.java:110]: Nitpick: we usually put `final` fields before the non-final ones.

**Review by tmysik — APPROVED** [review]
Not a small change, perfectly done! Thanks a lot!

File: php/php.editor/src/org/netbeans/modules/php/editor/actions/UsedNamesCollector.java (hunk 13 above)
Comment (rossluk) [reply php/php.editor/src/org/netbeans/modules/php/editor/actions/UsedNamesCollector.java]: Fixed.
